driver/k8s/logicalvolume_service.go

```diff
@@ -108,25 +109,49 @@ func NewLogicalVolumeService(mgr manager.Manager) (*LogicalVolumeService, error)
 }
 
 // CreateVolume creates volume
-func (s *LogicalVolumeService) CreateVolume(ctx context.Context, node, dc, name string, requestGb int64) (string, error) {
-	logger.Info("k8s.CreateVolume called", "name", name, "node", node, "size_gb", requestGb)
+func (s *LogicalVolumeService) CreateVolume(ctx context.Context, node, dc, name, sourceName string, requestGb int64) (string, error) {
+	logger.Info("k8s.CreateVolume called", "name", name, "node", node, "size_gb", requestGb, "sourceName", sourceName)
 	s.mu.Lock()
 	defer s.mu.Unlock()
+	var lv *topolvmv1.LogicalVolume
+	// if the create volume request has no source, proceed with regular lv creation.
+	if sourceName == "" {
+		lv = &topolvmv1.LogicalVolume{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "LogicalVolume",
+				APIVersion: "topolvm.cybozu.com/v1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name: name,
+			},
+			Spec: topolvmv1.LogicalVolumeSpec{
+				Name:        name,
+				NodeName:    node,
+				DeviceClass: dc,
+				Size:        *resource.NewQuantity(requestGb<<30, resource.BinarySI),
+			},
+		}
 
-	lv := &topolvmv1.LogicalVolume{
-		TypeMeta: metav1.TypeMeta{
-			Kind:       "LogicalVolume",
-			APIVersion: "topolvm.cybozu.com/v1",
-		},
-		ObjectMeta: metav1.ObjectMeta{
-			Name: name,
-		},
-		Spec: topolvmv1.LogicalVolumeSpec{
-			Name:        name,
-			NodeName:    node,
-			DeviceClass: dc,
-			Size:        *resource.NewQuantity(requestGb<<30, resource.BinarySI),
-		},
+	} else {
+		// On the other hand, if a volume has a datasource, create a thin snapshot of the source volume with READ-WRITE access.
+		lv = &topolvmv1.LogicalVolume{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "LogicalVolume",
+				APIVersion: "topolvm.cybozu.com/v1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name: name,
+			},
+			Spec: topolvmv1.LogicalVolumeSpec{
+				Name:        name,
+				NodeName:    node,
+				DeviceClass: dc,
+				Size:        *resource.NewQuantity(requestGb<<30, resource.BinarySI),
+				Type:        "thin-snapshot",
+				Source:      sourceName,
```

Comment:
Is it wrong?
```suggestion
				Size:        *resource.NewQuantity(requestGb<<30, resource.BinarySI),
				Source:      sourceName,
```

Reply:
https://github.com/topolvm/topolvm/pull/463#discussion_r887608132